enhancements/authentication/improved-login-workflow.md

```diff
@@ -0,0 +1,255 @@
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The current CLI login process involves multiple interactive steps where the user has to enter credentials multiple times
+in the browser. This process can be simplified so that the `oc` tool can be used to login directly without having to
+first login to the cluster console, the OAuth server and finally copying the login command. The `oc` tool can be
+modified to act as a public client to the Openshift OAuth server and received an access token
+through [Authorization Grant](https://datatracker.ietf.org/doc/html/rfc6749#section-1.3) flow
+and [PKCE](https://datatracker.ietf.org/doc/html/rfc7636).
+
+## Motivation
+
+Currently, when logging in with the `oc` client a user has to perform the following steps:
+
+1. Open the console in the browser.
+2. Login to console with the configured identity provider.
+3. Click on the **"Copy Login Command"** button which redirects the page to the OAuth application.
+4. Perform login again like in step 2.
+5. Copy the displayed login command.
+6. Run the copied command to configure the client to configure the client and access the cluster.
+
+This process can be simplified by using OAuth Authorization Grants for configuring `oc` without having to perform a
+login multiple times and also without copy-pasting the login command with the token.
+
+### Goals
+
+Provide a single-step login in `oc` for identity providers that don't post HTTP challenges.
+
+### Non-Goals
+
+None
+
+## Proposal
+
+### User Stories
+
+__Story 1__ As a user I want to be able to login to the cluster with `oc` and authenticating/providing credentials at
+most once.
+
+__Story 2__ As a user I want the login without having to copy and paste credentials in plain-text.
+
+### API Extensions
+
+None
+
+### Implementation Details [optional]
+
+#### _oc login_ command
+
+The oc login command will have a new flag `--auto` which indicates to `oc` that it should use the Oauth2 Authorization
+Flow to login. Currently, when a user enters `oc login $CLUSTER_API_URL`, `oc` attempts to be login on the cluster if
+the cluster has identity provider which
+support [challenge based authentication](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate).
+The new login flow cannot be a fallback if challenge based authentication is unavailable because the user might
+explicitly want to use it to login through code flow with PKCE. For this reason a new flag which indicates to `oc` to
+start the code flow is required.
+
+#### OAuth server
+
+The internal OAuth server uses [OSIN](https://github.com/openshift/osin) as a dependency to provide OAuth2 Authorized
```

Comment:
does "Authorized Grant Flow" mean "Authorization Code Grant" flow here or is it just "OAuth2-authorized flows"?